search/localsearch_nqueens.cpp

```diff
@@ -0,0 +1,184 @@
+ * @file
+ * @brief [Heuristic local Search](https://en.wikipedia.org/wiki/Local_search_(optimization)) algorithm.
+ * @details
+ * The LS algorithm is the basis of many methods used for optimization problems.
+ * Its system is based on the search for neighborhoods that obtain a better solution.
+ * This exercise solves the n queens problem [https://en.wikipedia.org/wiki/Eight_queens_puzzle].
+ * The rule of neighborhood applied in this file is first.improvement.
+ * @author [dhernandezl](https://github.com/dhernandezl)
+ */
+
+#include <iostream>
+#include <vector>
+#include <algorithm>
+#include <ctime>
+
+#define NREINAS 8 //The number of queens on the table is the same as the dimension of the chessboard.
+				  // NREINAS 4; NREINAS 8; NREINAS 16
+/**
+ * function: initial_solution
+ * Initialize a new solution, generating
+ * random numbers with the maximum height of the board.
+ * @returns a vector that will be the starting solution.
+ */
+std::vector<int> initial_solution() {
+	std::vector<int> solution;
+	srand(time(nullptr)); //reset srand
+	for (size_t i = 0; i < NREINAS; i++)
+	{
+		solution.push_back(rand() % NREINAS + 1);
+	}
+	return solution;
+}
+
+/**
+ * function: cost
+ * Calculate the cost of the solution vector,
+ * the optimal cost is h = 1. Read the 8 queens problem.
+ * @param [in] solution_vector is a vector with the
+ * column number where the queens are located on the board.
+ * @returns the cost of the solution vector.
+ */
+int cost(std::vector<int> solution_vector) {
+	int h = 0;
+	std::vector<int> aux_sol;
+	//sum column
+	for (size_t i = 0; i < NREINAS; i++)
+	{
+		for (size_t j = i + 1; j < NREINAS; j++)
+		{
+			if (solution_vector[i] == solution_vector[j])
+			{
+				h++;
+				break;
+			}
+		}
+	}
+	//sum diagonal
+	for (size_t i = 0; i < NREINAS; i++)
+	{
+		aux_sol.push_back(solution_vector[i] - (i + 1));
+	}
+	for (size_t i = 0; i < NREINAS; i++)
+	{
+		for (size_t j = i + 1; j < NREINAS; j++)
+		{
+			if (aux_sol[i] == aux_sol[j])
+			{
+				h++;
+				break;
+			}
+		}
+	}
+	aux_sol.clear();
+	//sum inverse diagonal
+	for (size_t i = 0; i < NREINAS; i++)
+	{
+		aux_sol.push_back(solution_vector[i] + (i + 1));
+	}
+	for (size_t i = 0; i < NREINAS; i++)
+	{
+		for (size_t j = i + 1; j < NREINAS; j++)
+		{
+			if (aux_sol[i] == aux_sol[j])
+			{
+				h++;
+				break;
+			}
+		}
+	}
+	return h;
+}
+
+/**
+ * function: neighbour
+ * Generate a neighbor based on the current solution of
+ * the problem. Read about neighborhoods in local search.
+ * @param [in] solution_vector is a vector with the
+ * column number where the queens are located on the board.
+ * @returns a new vector similar to the one passed by parameter.
+ */
+std::vector<int> neighbour(std::vector<int> solution_vector) {
+	srand(time(nullptr)); //reset srand
+	int i = rand() % NREINAS + 1;
+	srand(time(nullptr)); //reset srand
+	int mov = (rand() % NREINAS + 1) - i;
+	solution_vector[i - 1] = (mov < 0) ? mov *= -1: (mov == 0) ? mov = 1 : mov;

```

Comment:
This part is very unclear and deserves at least a brief explanation in a comment.
After reading it carefully, it occurs to me that you're basically generating a new value for a randomly chosen element of the `solution_vector`, however, the distribution for the values is not uniform. For small and large `i` the distribution of the final value is going to be fairly uniform, but for `i` in the middle (e.g. `NREINAS / 2`) the chosen value is only going to be in a range from `1` to `NREINAS / 2`, I don't believe it's desirable.